src/query/mlt/mlt.rs

```diff
@@ -0,0 +1,387 @@
+                )
+            })
+            .collect::<Vec<_>>();
+        self.retrieve_terms_from_doc_fields(searcher, &field_to_field_values)
+    }
+
+    /// Finds terms for a more-like-this query.
+    /// field_to_field_values is a mapping from field to possible values of taht field.
+    fn retrieve_terms_from_doc_fields(
+        &self,
+        searcher: &Searcher,
+        field_to_field_values: &[(Field, Vec<FieldValue>)],
+    ) -> Result<Vec<ScoreTerm>> {
+        if field_to_field_values.is_empty() {
+            return Err(TantivyError::InvalidArgument("Cannot create more like this query on empty field values. The document may not have stored fields".to_string()));
+        }
+
+        let mut field_to_term_freq_map = HashMap::new();
+        for (field, field_values) in field_to_field_values {
+            self.add_term_frequencies(searcher, *field, field_values, &mut field_to_term_freq_map)?;
+        }
+        self.create_score_term(searcher, field_to_term_freq_map)
+    }
+
+    /// Computes the frequency of values for a field while updating the term frequencies
+    /// Note: A FieldValue can be made up of multiple terms.
+    /// We are interested in extracting terms within FieldValue
+    fn add_term_frequencies(
+        &self,
+        searcher: &Searcher,
+        field: Field,
+        field_values: &[FieldValue],
+        term_frequencies: &mut HashMap<Term, usize>,
+    ) -> Result<()> {
+        let schema = searcher.schema();
+        let tokenizer_manager = searcher.index().tokenizers();
+
+        let field_entry = schema.get_field_entry(field);
+        if !field_entry.is_indexed() {
+            return Ok(());
+        }
+
+        // extract the raw value, possibly tokenizing & filtering to update the term frequency map
+        match field_entry.field_type() {
+            FieldType::HierarchicalFacet(_) => {
+                let facets: Vec<&str> = field_values
+                    .iter()
+                    .map(|field_value| match *field_value.value() {
+                        Value::Facet(ref facet) => Ok(facet.encoded_str()),
+                        _ => Err(TantivyError::InvalidArgument(
+                            "invalid field value".to_string(),
+                        )),
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+                for fake_str in facets {
+                    FacetTokenizer.token_stream(fake_str).process(&mut |token| {
+                        if self.is_noise_word(token.text.clone()) {
+                            let term = Term::from_field_text(field, &token.text);
+                            *term_frequencies.entry(term).or_insert(0) += 1;
+                        }
+                    });
+                }
+            }
+            FieldType::Str(text_options) => {
+                let mut token_streams: Vec<BoxTokenStream> = vec![];
+                let mut offsets = vec![];
```

Comment:
what are the offsets for!???

Reply:
Thanks for the eyes. lot of issue with the borrow checker so i was thinking of another strategy. but resultved it. I think there still one comment not needed there. 
Just wondering whay `clippy` did not spot this